fedbiomed/researcher/experiment.py

```diff
@@ -1643,6 +1647,36 @@ def run_once(self, increase: bool = False, test_after: bool = False) -> int:
 
         return 1
 
+    def _process_optim_aux_var(
+        self,
+    ) -> None:
+        """Process any Optimizer auxiliary variables received during a round.
+
+        Raises:
+            FedbiomedExperimentError: if auxiliary variables were received,
+                but `agg_optimizer` is None and thus cannot process them.
+            FedbiomedOptimizerError: if the received auxiliary variables do
+                not match the expectations of the `agg_optimizer` Optimizer.
+        """
+        # Restructure the received auxiliary variables (if any).
```

Comment:
I agree on the principle. I think part of it is inevitable with the current partition of the workload between `Job` and `Experiment`: `Job` is in charge of collecting and sending information, then gathering replies, while `Experiment` is in charge of conducting their post-processing (aggregation, optimization and application of updates); but I also think that I can improve my implementation to better stick to that partition.

I think it would make sense to:
- move the re-organization of received auxiliary variables to `Job`, so that `Experiment._process_optim_aux_var` can be shrinked to merely passing the reply-contained aux vars to the Optimizer (and handling incoherence issues).
- change `Job.upload_agg_optimizer_aux_var` signature so that instead of `agg_optimizer: Optional[Optimizer]` it expects `optim_aux_var: Optional[Dict[str, Dict[str, Any]]]`, and have `Experiment` perform the explicit call to `Optimizer.get_aux`

This way the structuration of information (for upload and at download) would all be part of `Job`, while the collection and processing at the level of the `Optimizer` would both be performed directly by `Experiment`.

Would that suit you?

Reply:
Update: I implemented the changes suggested above based on your remark. Let me know if that suits you :)

Reply:
> This way the structuration of information (for upload and at download) would all be part of Job, while the collection and processing at the level of the Optimizer would both be performed directly by Experiment.

> Would that suit you?

Yes this is what I think we should do!

Reply:
*Update**: I checked and I think and the way you re-design things is way better in term of responsibility! 
Only one small detail: ``get_received_optimizer_aux_var` could also be in `Experiment` (but I guess you did that in order to avoid appending it to Experiment-that-really-need-some-refactoring :) )

Reply:
This is merely a suggestion, please resolve the conversation if you think it is doesnot make sense

Reply:
> Only one small detail: `get_received_optimizer_aux_var` could also be in `Experiment`

Well, in fact it _was_ in `Experiment` (as part of `_process_optim_aux_var`) prior to the refactoring...

I think I like it better where it is, because this way the bits of code that handle the re-structuring of aux var in both directions (researcher <-> nodes) are grouped together within `Job`. But if you disagree, I can revert that part of the changes and change the unit tests, so let me know.

Reply:
Ok then it makes even more sense to keep it in `Job`!
Ok let's keep it in `Job` then!